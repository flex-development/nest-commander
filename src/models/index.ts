/**
 * @file Entry Point - Models
 * @module nest-commander/models
 */

export { default as ProgramOptions } from './options-program.model'
