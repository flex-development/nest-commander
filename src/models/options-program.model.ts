/**
 * @file Models - ProgramOptions
 * @module nest-commander/models/ProgramOptions
 */

import type { VersionOptions } from '#src/interfaces'
import type { DoneFn, ErrorFn, ExitFn } from '#src/types'
import { fallback, noop } from '@flex-development/tutils'

/**
 * CLI program options.
 *
 * @class
 */
class ProgramOptions {
  /**
   * Alter parsing of short flags with optional values.
   *
   * @see https://github.com/tj/commander.js#options
   *
   * @default false
   *
   * @public
   * @instance
   * @member {boolean?} combine
   */
  public combine?: boolean

  /**
   * Callback ran after command-line arguments are parsed.
   *
   * @see {@linkcode DoneFn}
   *
   * @default noop
   *
   * @public
   * @instance
   * @member {DoneFn<any>?} done
   */
  public done?: DoneFn<any>

  /**
   * Handles an error thrown by a command.
   *
   * @see {@linkcode ErrorFn}
   *
   * @default console.error
   *
   * @public
   * @instance
   * @member {ErrorFn<any>?} error
   */
  public error?: ErrorFn<any>

  /**
   * Allow excess command-arguments on the command line.
   *
   * @see https://github.com/tj/commander.js#action-handler
   *
   * @default false
   *
   * @public
   * @instance
   * @member {boolean?} excess
   */
  public excess?: boolean

  /**
   * Callback to use as replacement for calling {@linkcode process.exit}.
   *
   * @see {@linkcode ExitFn}
   *
   * @default e => process.exit(e.exitCode)
   *
   * @public
   * @instance
   * @member {ExitFn?} exit
   */
  public exit?: ExitFn

  /**
   * Only process options that come before command arguments.
   *
   * @see https://github.com/tj/commander.js#parsing-configuration
   *
   * @default false
   *
   * @public
   * @instance
   * @member {boolean?} passthrough
   */
  public passthrough?: boolean

  /**
   * Only look for program options before subcommands.
   *
   * @see https://github.com/tj/commander.js#parsing-configuration
   *
   * @default true
   *
   * @public
   * @instance
   * @member {boolean?} positional
   */
  public positional?: boolean

  /**
   * Allow unknown options on the command line.
   *
   * @see https://github.com/tj/commander.js#parsing-configuration
   *
   * @default false
   *
   * @public
   * @instance
   * @member {boolean?}
   */
  public unknown?: boolean

  /**
   * Program version options.
   *
   * @see {@linkcode VersionOptions}
   *
   * @default ''
   *
   * @public
   * @instance
   * @member {(VersionOptions | VersionOptions['version'])?} version
   */
  public version?: VersionOptions | VersionOptions['version']

  /**
   * Creates new program options object.
   *
   * @param {Partial<ProgramOptions>} [options={}] - CLI program options
   */
  constructor(options: Partial<ProgramOptions> = {}) {
    const {
      combine,
      done,
      error,
      excess,
      exit,
      passthrough,
      positional,
      unknown,
      version
    } = options

    this.combine = fallback(combine, false)
    this.done = fallback(done, noop)
    this.error = fallback(error, console.error)
    this.excess = fallback(excess, false)
    this.exit = fallback(exit, e => process.exit(e.exitCode))
    this.passthrough = fallback(passthrough, false)
    this.positional = fallback(positional, true)
    this.unknown = fallback(unknown, false)
    this.version = fallback(version, '')
  }
}

export default ProgramOptions
