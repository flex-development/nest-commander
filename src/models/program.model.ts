/**
 * @file Models - Program
 * @module nest-commander/models/Program
 */

import { Command, type ParseOptions } from '#src/commander'
import {
  cast,
  fallback,
  get,
  includes,
  keys,
  type Nilable,
  type Optional
} from '@flex-development/tutils'
import ProgramOptions from './options-program.model'

/**
 * CLI program model.
 *
 * @see {@linkcode Command}
 *
 * @class
 * @extends {Command}
 */
class Program extends Command {
  /**
   * Program configuration options.
   *
   * @see {@linkcode ProgramOptions}
   *
   * @public
   * @readonly
   * @instance
   * @member {Readonly<Required<ProgramOptions>>} config
   */
  public readonly config: Readonly<Required<ProgramOptions>>

  /**
   * Parent command.
   *
   * @public
   * @override
   * @readonly
   * @instance
   * @member {null} parent
   */
  public override readonly parent: null

  /**
   * Creates a new CLI program instance.
   *
   * @see {@linkcode ProgramOptions}
   *
   * @param {ProgramOptions?} [options] - Program options
   * @param {string?} [options.name] - Program name
   */
  constructor(options?: ProgramOptions & { name?: string }) {
    super(get(options, 'name'))
    this.config = cast(Object.freeze(new ProgramOptions(options)))
    this.parent = null

    // configure argument and option parsing
    this.allowExcessArguments(this.config.excess)
    this.allowUnknownOption(this.config.unknown)
    this.combineFlagAndOptionalValue(this.config.combine)
    this.enablePositionalOptions(this.config.positional)
    this.passThroughOptions(this.config.passthrough)

    // configure exit override
    this.exitOverride(this.config.exit)

    // configure program version
    if (keys(this.config.version).length) {
      this.version(
        get(this.config.version, 'version', cast<string>(this.config.version)),
        get(this.config.version, 'flags'),
        get(this.config.version, 'description')
      )
    }
  }

  /**
   * Copy common settings to `command`.
   *
   * @public
   *
   * @param {Command} command - Command to copy settings to
   * @return {typeof command} Updated `command`
   */
  public enforceSettings(command: Command): typeof command {
    command.allowUnknownOption(this.config.unknown)
    command.passThroughOptions(this.config.passthrough)
    return command.copyInheritedSettings(this)
  }

  /**
   * Find a command by name or alias.
   *
   * @public
   *
   * @param {Nilable<string>} cmd - Command name or alias
   * @param {Command?} [parent=this] - Parent command
   * @return {Optional<Command>} Command named `name` or `undefined`
   */
  public findCommand(
    cmd: Nilable<string>,
    parent: Command = this
  ): Optional<Command> {
    return parent.commands.find(command => {
      return command.name() === cmd || includes(command.aliases(), cmd)
    })
  }

  /**
   * Parse `argv`, setting options and invoking commands when defined, as well
   * as calling the `done` ({@linkcode ProgramOptions.done}) callback when
   * parsing complete.
   *
   * The default expectation is that arguments are from `node` and have the
   * application as `argv[0]` and current script as `argv[1]`, with any user
   * parameters thereafter. This behavior can be changed with `options.from`.
   *
   * @see https://github.com/tj/commander.js#parse-and-parseasync
   *
   * @public
   * @override
   * @async
   *
   * @param {ReadonlyArray<string>?} [argv=process.argv] - CLI arguments
   * @param {ParseOptions?} [options] - CLI argument parsing options
   * @param {ParseOptions['from']} options.from - CLI arguments source
   * @return {Promise<this>} CLI program instance
   */
  public override async parseAsync(
    argv: readonly string[] = process.argv,
    options?: ParseOptions
  ): Promise<this> {
    // parse command-line arguments
    await super.parseAsync(argv, options)

    /**
     * Command that was run.
     *
     * @const {Command} command
     */
    const command: Command = fallback(this.findCommand(this.args[0]), this)

    // run post-parse callback
    await this.config.done(command.args, command.optsWithGlobals(), command)

    return this
  }
}

export default Program
