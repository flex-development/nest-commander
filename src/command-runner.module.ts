/**
 * @file CommandRunnerModule
 * @module nest-commander/CommandRunnerModule
 */

import { sift, type Class } from '@flex-development/tutils'
import { DiscoveryModule } from '@golevelup/nestjs-discovery'
import { Module, type DynamicModule } from '@nestjs/common'
import { Program, ProgramOptions } from './models'
import {
  CliUtilityService,
  CommandRunnerService,
  HelpService
} from './providers'

/**
 * Command runner module.
 *
 * @class
 */
@Module({})
class CommandRunnerModule {
  /**
   * Register a global command runner module.
   *
   * @public
   * @static
   *
   * @param {ProgramOptions} options - CLI program options
   * @param {(Class<any> | DynamicModule)?} [module] - Module to import
   * @return {DynamicModule} Global command runner module
   */
  public static register(
    options: ProgramOptions,
    module?: Class<any> | DynamicModule
  ): DynamicModule {
    return {
      exports: [CliUtilityService, HelpService, Program],
      global: true,
      imports: sift([DiscoveryModule, module]),
      module: CommandRunnerModule,
      providers: [
        {
          provide: ProgramOptions,
          useValue: new ProgramOptions(options)
        },
        {
          inject: [ProgramOptions, HelpService],
          provide: Program,
          useFactory(options: ProgramOptions, help: HelpService): Program {
            return new Program(options).configureHelp(help)
          }
        },
        CliUtilityService,
        CommandRunnerService,
        HelpService
      ]
    }
  }
}

export default CommandRunnerModule
