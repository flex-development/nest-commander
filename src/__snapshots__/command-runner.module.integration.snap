// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`integration:CommandRunnerModule > program with multiple commands > should output help text when requested 1`] = `
"Description
  convert a list of a string

Usage
  $ string-utils join [options] [strings...]

Examples
  $ string-utils join hello world --separator=.

Options
  -s, --separator <choice>  list item separator (default: \\",\\")
  -h, --help                display help for command
"
`;

exports[`integration:CommandRunnerModule > program with root command > should output help text when requested 1`] = `
"Description
  Toggle type fields in package.json files

Usage
  $ toggle-pkg-type [options] [cmd]

Examples
  $ toggle-pkg-type
  $ toggle-pkg-type on
  $ toggle-pkg-type off
  $ npm_package_json=/path/to/directory/or/manifest toggle-pkg-type

Arguments
  cmd              Toggle command (choices: \\"off\\", \\"on\\")

Options
  -i, --id <path>  module id of package directory or manifest (default: \\".\\", env: npm_package_json)
  -h, --help       display help for command
  -v, --version    print version number
"
`;

exports[`integration:CommandRunnerModule > program with subcommands > should output help text when requested 1`] = `
"Description
  get proper timezone abbreviation or timezone offset

Usage
  $ dateformat timezone|tz [options] [date]

Examples
  $ dateformat timezone
  $ dateformat tz
  $ dateformat timezone 2023-08-21T21:45:48.850Z
  $ dateformat tz 2023-08-23T00:46:32.292Z

Arguments
  date        date string

Options
  -h, --help  display help for command
"
`;
